File: src/main/java/com/janrain/backplane/server2/Token.java
/*
 * Copyright 2012 Janrain, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.janrain.backplane.server2;

import com.janrain.backplane.common.BackplaneServerException;
import com.janrain.backplane.common.DateTimeUtils;
import com.janrain.backplane.common.ExternalizableCore;
import com.janrain.backplane.common.RandomUtils;
import com.janrain.backplane.server2.dao.BP2DAOs;
import com.janrain.backplane.server2.oauth2.OAuth2;
import com.janrain.backplane.server2.oauth2.TokenException;
import com.janrain.backplane.servlet.InvalidRequestException;
import com.janrain.commons.message.AbstractMessage;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;
import com.janrain.commons.util.Pair;
import org.apache.commons.lang.StringUtils;
import org.apache.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.text.ParseException;
import java.util.*;

import static com.janrain.backplane.server2.oauth2.OAuth2.*;

/**
 * @author Tom Raney, Johnny Bufu
 */

public class Token extends ExternalizableCore {

    /**
     * Empty default constructor for AWS to use.
     * Don't call directly.
     */
    public Token() {}

    @Override
    public String getIdValue() {
        return get(TokenField.ID);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(TokenField.class);
    }

    @Override
    public void validate() throws MessageException {
        super.validate();
        if (getType().isPrivileged()) {
            AbstractMessage.validateNotBlank(TokenField.ISSUED_TO_CLIENT_ID.getFieldName(), get(TokenField.ISSUED_TO_CLIENT_ID));
            AbstractMessage.validateNotBlank(TokenField.CLIENT_SOURCE_URL.getFieldName(), get(TokenField.CLIENT_SOURCE_URL));
            AbstractMessage.validateNotBlank(TokenField.BACKING_GRANTS.getFieldName(), get(Token.TokenField.BACKING_GRANTS));
        } else {
            Scope anonScope = getScope();
            LinkedHashSet<String> buses = anonScope.getScopeMap().get(BackplaneMessage.Field.BUS);
            LinkedHashSet<String> channels = anonScope.getScopeMap().get(BackplaneMessage.Field.CHANNEL);
            if (buses == null || buses.size() > 1 || channels == null || channels.size() > 1) {
                throw new MessageException("invalid scope for anonymous token, must have exactly one bus and one channel specified: " + anonScope);
            }
        }
    }

    public GrantType getType() {
        try {
            return GrantType.valueOf(this.get(TokenField.TYPE));
        } catch (IllegalArgumentException e) {
            throw new IllegalStateException("Invalid GrantType on for TokenField.Type, should have been validated on token creation: " + this.get(TokenField.TYPE));
        }
    }

    public Scope getScope()  {
        try {
            return new Scope(this.get(TokenField.SCOPE));
        } catch (TokenException e) {
            throw new IllegalStateException("Invalid scope on get(), should have been validated on token creation: " + this.get(TokenField.SCOPE));
        }
    }

    /**
     * @return the token's expiration date, or null if the token never expires
     */
    public Date getExpirationDate() {
        String value = this.get(TokenField.EXPIRES);
        try {
            return StringUtils.isNotEmpty(value) ? DateTimeUtils.ISO8601.get().parse(value) : null;
        } catch (ParseException e) {
            throw new IllegalStateException("Invalid ISO8601 date for TokenField.EXPIRES, should have been validated on token creation: " + value);
        } catch (NumberFormatException nfe) {
            logger.error("Error parsing token date: " + value, nfe);
            throw new IllegalStateException("Invalid ISO8601 date for TokenField.EXPIRES, should have been validated on token creation: " + value);
        }
    }

    public String getScopeString() {
        return get(TokenField.SCOPE);
    }

    public boolean isExpired() {
        Date expires = getExpirationDate();
        return expires != null && new Date().getTime() > expires.getTime();
    }

    public static boolean looksLikeOurToken(String tokenString) {
        GrantType grantType = GrantType.fromTokenString(tokenString);
        if (grantType == null) return false;
        String tokenNoPrefix = tokenString.substring(grantType.getTokenPrefix().length());
        return tokenNoPrefix.length() == TOKEN_LENGTH;
    }

    public static @NotNull Token fromRequest(HttpServletRequest request, String tokenString, String authorizationHeader) throws TokenException {
        
        Pair<String, EnumSet<TokenSource>> tokenAndSource = extractToken(request.getQueryString(), tokenString, authorizationHeader);

        if (! Token.looksLikeOurToken(tokenAndSource.getLeft())) {
            throw new TokenException("invalid token", HttpServletResponse.SC_FORBIDDEN);
        }

        Token token;
        try {
            token = BP2DAOs.getTokenDao().get(tokenAndSource.getLeft());
        } catch (BackplaneServerException e) {
            logger.error("Error looking up token: " + tokenAndSource.getLeft() , e);
            throw new TokenException(OAuth2.OAUTH2_TOKEN_SERVER_ERROR, "error loading token", HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
        }

        if (token == null) {
            logger.warn("token not found: " + tokenAndSource.getLeft());
            throw new TokenException("invalid token", HttpServletResponse.SC_FORBIDDEN);
        }

        if (token.isExpired()) {
            throw new TokenException("expired token", HttpServletResponse.SC_FORBIDDEN);
        }

        token.checkAllowedSources(tokenAndSource.getRight());

        return token;
    }

    public Map<String, Object> response(final String refreshToken) {
        final Date expires = getExpirationDate();
        return new LinkedHashMap<String, Object>() {{
            put(OAUTH2_TOKEN_TYPE_PARAM_NAME, OAUTH2_TOKEN_TYPE_BEARER);
            put(OAUTH2_ACCESS_TOKEN_PARAM_NAME, getIdValue());
            if (expires != null) put(OAUTH2_TOKEN_RESPONSE_EXPIRES, (expires.getTime() - new Date().getTime()) / 1000);
            put(OAUTH2_SCOPE_PARAM_NAME, getScopeString());
            if (refreshToken != null) put(OAUTH2_REFRESH_TOKEN_PARAM_NAME, refreshToken);
        }};
    }

    public @NotNull List<String> getBackingGrants() {
        String grants = get(TokenField.BACKING_GRANTS);
        return StringUtils.isNotEmpty(grants) ? Arrays.asList(grants.split(GRANTS_SEPARATOR)) : new ArrayList<String>();
    }

    public static enum TokenField implements MessageField {

        ID("id", true),

        TYPE("type", true) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                try {
                    GrantType.valueOf(value);
                } catch (IllegalArgumentException e) {
                    throw new MessageException("Invalid token type: " + value);
                }
            }
        },

        EXPIRES("expires", false) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                try {
                    if (StringUtils.isNotEmpty(value)) {
                        DateTimeUtils.ISO8601.get().parse(value);
                    }
                } catch (ParseException e) {
                    throw new MessageException("Invalid token expiration date: " + value, e);
                }
            }
        },

        SCOPE("scope", true) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                try {
                    new Scope(value);
                } catch (TokenException e) {
                    throw new InvalidRequestException("Invalid scope: " + value);
                }
            }
        },

        ISSUED_TO_CLIENT_ID("issued_to_client", false),
        
        CLIENT_SOURCE_URL("client_source_url", false) {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                if (StringUtils.isNotEmpty(value)) {
                    AbstractMessage.validateUrl(getFieldName(), value);
                }
            }
        },

        BACKING_GRANTS("backing_grants", false);

        @Override
        public String getFieldName() {
            return fieldName;
        }

        @Override
        public boolean isRequired() {
            return required;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(getFieldName(), value);
        }

        // - PRIVATE

        private String fieldName;
        private boolean required = true;

        private TokenField(String fieldName, boolean required) {
            this.fieldName = fieldName;
            this.required = required;
        }
    }

    public static class Builder {

        public Builder(GrantType type, String scope) {
            this.type = type;
            data.put(TokenField.TYPE.getFieldName(), type.toString());
            data.put(TokenField.SCOPE.getFieldName(), scope);
        }

        public Builder expires(Date expires) {
            data.put(TokenField.EXPIRES.getFieldName(), DateTimeUtils.ISO8601.get().format(expires));
            return this;
        }
        
        public Builder issuedToClient(String clientId) {
            data.put(TokenField.ISSUED_TO_CLIENT_ID.getFieldName(), clientId);
            return this;
        }
        
        public Builder clientSourceUrl(String clientSourceUrl) {
            data.put(TokenField.CLIENT_SOURCE_URL.getFieldName(), clientSourceUrl);
            return this;
        }

        public Builder grants(List<String> grants) {
            data.put(TokenField.BACKING_GRANTS.getFieldName(),
                     org.springframework.util.StringUtils.collectionToDelimitedString(grants, GRANTS_SEPARATOR));
            return this;
        }
        
        public Token buildToken() throws MessageException {
            String id = type.getTokenPrefix() + RandomUtils.randomString(TOKEN_LENGTH);
            data.put(TokenField.ID.getFieldName(), id);
            return new Token(id, data);
        }
        
        // - PRIVATE
        
        private final Map<String,String> data = new HashMap<String, String>();
        private final GrantType type;
    }
    
    // - PRIVATE

    private static final long serialVersionUID = -5883124096459804032L;
    
    private static final Logger logger = Logger.getLogger(Token.class);

    private static final int TOKEN_LENGTH = 20;

    private static final String GRANTS_SEPARATOR = " ";

    private Token(String id, Map<String,String> data) throws MessageException {
        super.init(id, data);
        logger.debug("created token: " + this.toString());
    }

    private static Pair<String,EnumSet<TokenSource>> extractToken(String queryString, String requestParam, String authHeader) {
        String token = null;
        EnumSet<TokenSource> foundIn = EnumSet.noneOf(TokenSource.class);

        if (StringUtils.isNotEmpty(queryString)) {
            Map<String,String> queryParamsMap = new HashMap<String, String>();
            for(String queryParamPair : Arrays.asList(queryString.split("&"))) {
                String[] nameVal = queryParamPair.split("=", 2);
                queryParamsMap.put(nameVal[0], nameVal.length >1 ? nameVal[1] : null);
            }
            if(queryParamsMap.containsKey(OAUTH2_ACCESS_TOKEN_PARAM_NAME)) {
                token = queryParamsMap.get(OAUTH2_ACCESS_TOKEN_PARAM_NAME);
                foundIn.add(TokenSource.QUERYPARAM);
            }
        }

        if ( ! foundIn.contains(TokenSource.QUERYPARAM) && requestParam != null ) {
            // query parameter will mask body requestParam extracted by spring with @RequestParameter
            token = requestParam;
            foundIn.add(TokenSource.POSTBODY);
        }

        int tokenTypeLength = OAUTH2_TOKEN_TYPE_BEARER.length();
        if (authHeader != null && authHeader.startsWith(OAUTH2_TOKEN_TYPE_BEARER) && authHeader.length() > tokenTypeLength + 1) {
            token = authHeader.substring(tokenTypeLength + 1);
            foundIn.add(TokenSource.AUTHHEADER);
        }

        return new Pair<String, EnumSet<TokenSource>>(token, foundIn);
    }

    private void checkAllowedSources(Collection<TokenSource> tokenFoundIn) throws TokenException {
        if (tokenFoundIn == null || tokenFoundIn.size() > 1) {
            throw new TokenException("exactly one token source allowed, found in: " + tokenFoundIn, HttpServletResponse.SC_FORBIDDEN);
        }

        for(TokenSource tokenSource : tokenFoundIn) {
            if (! getType().getTokenAllowedSources().contains(tokenSource)) {
                throw new TokenException("token source not allowed: " + tokenSource, HttpServletResponse.SC_FORBIDDEN);
            }
        }
    }
}

File: src/main/scala/com/janrain/backplane/server2/oauth2/model/Token.scala
package com.janrain.backplane.server2.oauth2.model

import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}
import com.janrain.backplane2.server.{Scope, GrantType}
import com.janrain.backplane.common.MessageException
import com.janrain.oauth2.{OAuth2, TokenException}
import com.janrain.oauth2.OAuth2._
import javax.servlet.http.{HttpServletResponse, HttpServletRequest}
import com.janrain.backplane.server2.TokenSource
import com.janrain.backplane.server2.dao.{BP2DAOs}
import com.janrain.backplane.dao.{LegacySupport, DaoException}
import com.janrain.util.{Utils, Loggable}
import scala.collection.JavaConversions._
import java.util.Date
import com.janrain.backplane.server2.model.Backplane2MessageFields

/**
 * @author Johnny Bufu
 */
class Token(data: Map[String,String]) extends Message(data, TokenFields.values) with LegacySupport[com.janrain.backplane2.server.Token] {

  def this(javaData: java.util.Map[String,String]) = this(LegacySupport.fromLegacy(javaData))

  def idField = TokenFields.ID

  def asLegacy = new com.janrain.backplane2.server.Token(mapAsJavaMap(this))

  def grantType: GrantType = get(TokenFields.TYPE).map(typeValue => {
    try {
      GrantType.valueOf(typeValue)
    } catch {
      case e: IllegalArgumentException =>
        throw new IllegalStateException("Invalid GrantType on for TokenFields.TYPE, should have been validated on grant creation: " + typeValue)
    }
  })
  .getOrElse( throw new IllegalStateException("Missing value for TokenFields.TYPE"))


  def scope: Scope = get(TokenFields.SCOPE).map(scopeValue => {
    try {
      new Scope(scopeValue)
    } catch {
      case e: TokenException =>
        throw new IllegalStateException("Invalid value on for TokenFields.SCOPE, should have been validated on grant creation: " + scopeValue)
    }
  })
  .getOrElse( throw new IllegalStateException("Missing value for TokenFields.SCOPE"))

  def allowedSources: Seq[TokenSource.EnumVal] = {
    val gt = grantType
    if (! gt.isPrivileged) TokenSource.values
    else if (gt.isRefresh) Seq(TokenSource.POSTBODY_REFRESH)
    else Seq(TokenSource.AUTHHEADER)
  }

  def backingGrants: List[String] = get(TokenFields.BACKING_GRANTS).map(_.split(Token.GRANTS_SEPARATOR)).toIterable.flatten.toList

  def response(refreshToken: String): java.util.Map[String, AnyRef] = {
    var response = new java.util.LinkedHashMap[String,AnyRef]()
    response.put(OAUTH2_TOKEN_TYPE_PARAM_NAME, OAUTH2_TOKEN_TYPE_BEARER)
    response.put(OAUTH2_ACCESS_TOKEN_PARAM_NAME, id)
    get(TokenFields.SCOPE).foreach( response.put(OAUTH2_SCOPE_PARAM_NAME, _) )
    get(TokenFields.EXPIRES).foreach { expValue =>
      val expireSeconds: Long = (Utils.ISO8601.parseDateTime(expValue).getMillis - new Date().getTime) / 1000
      response.put(OAUTH2_TOKEN_RESPONSE_EXPIRES, expireSeconds.toString )
    }
    if (refreshToken != null)
      response.put(OAUTH2_REFRESH_TOKEN_PARAM_NAME, refreshToken)
    response
  }

}

object Token extends Loggable {

  final val TOKEN_LENGTH: Int = 25
  final val LEGACY_TOKEN_LENGTH: Int = 20

  private def looksLikeOurToken(tokenString: String) = Option(GrantType.fromTokenString(tokenString))
    .map(gt => tokenString.substring(gt.getTokenPrefix.length))
    .exists(t => t.length == TOKEN_LENGTH || t.length == LEGACY_TOKEN_LENGTH)

  final val GRANTS_SEPARATOR = " "

  def fromRequest(request: HttpServletRequest): Option[Token] = {

    // multiple sources, multiple values per source considered
    val sourcesAndTokens = TokenSource.values.foldLeft( List[(TokenSource.EnumVal,List[String])]() ) {
      case (found, src) => src.extract(request, found.flatMap(_._2)) :: found
    }

    val foldAccInit: Option[(TokenSource.EnumVal, String)] = None
    val validSourceAndTokenString = sourcesAndTokens.foldLeft(foldAccInit) {
      case (found@Some(_), (nextSource, Nil)) => found
      case (found@Some(_), (nextSource, nextTokens)) => {
        throw new TokenException("token found in more than one source: %s and %s".format(found.get._1, nextSource))
      }
      case (None, (nextSource, Nil)) => None
      case (None, (nextSource, List(oneToken))) => Some((nextSource, oneToken))
      case (None, (nextSource, nextTokens)) => {
        throw new TokenException("multiple tokens found in " + nextSource)
      }
    }

    validSourceAndTokenString.map {
      case (source, tokenString) => {
        if (! Token.looksLikeOurToken(tokenString))
          throw new TokenException("invalid token: " + tokenString, HttpServletResponse.SC_FORBIDDEN)

        try {
          val token: Option[Token] = BP2DAOs.tokenDao.get(tokenString)

          if (token.exists(t => Message.isExpired(t.get(TokenFields.EXPIRES)))) {
            throw new TokenException("expired token: " + tokenString, HttpServletResponse.SC_FORBIDDEN)
          }

          if (! token.forall(t => t.allowedSources.contains(source))) {
            throw new TokenException("token source not allowed: " + source, HttpServletResponse.SC_FORBIDDEN)
          }

          if (token.exists(t => source.isRefresh != t.grantType.isRefresh)) {
            logger.warn("invalid token / source: %s / %s ".format(token, source))
            throw new TokenException("invalid token: " + tokenString, HttpServletResponse.SC_FORBIDDEN)
          }

          token
        } catch {
          case e: DaoException => {
            logger.error("Error looking up token: " + tokenString)
            throw new TokenException(OAuth2.OAUTH2_TOKEN_SERVER_ERROR, "error looking up token", HttpServletResponse.SC_INTERNAL_SERVER_ERROR)
          }
        }

      }
    }.flatten
  }
}

object TokenFields extends MessageFieldEnum {

  type TokenField = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
  }

  val ID = new TokenField { def name = "id" }

  val TYPE = new TokenField { def name = "type"
    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      fieldValue.foreach(grantValue => {
        try {
          GrantType.valueOf(grantValue)
        } catch {
          case e: IllegalArgumentException =>
            throw new MessageException("Invalid GrantType on for TokenField.Type " + grantValue)
        }
      })
    }
  }

  val EXPIRES = new TokenField { def name = "expires"
    override def required = false
    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      if (! wholeMessage.asInstanceOf[Token].grantType.isRefresh)
        validateRequired(fieldValue)
      validateIso8601Date(fieldValue)
    }
  }

  val SCOPE = new TokenField { def name = "scope"

    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      fieldValue.foreach(scopeValue => {
        try {
          val scope = new Scope(scopeValue)
          if (! wholeMessage.asInstanceOf[Token].grantType.isPrivileged) {
            val buses = scope.getScopeFieldValues(Backplane2MessageFields.BUS)
            val channels = scope.getScopeFieldValues(Backplane2MessageFields.CHANNEL)
            if (buses == null || buses.size > 1 || channels == null || channels.size > 1) {
              throw new MessageException("invalid scope for anonymous token, must have exactly one bus and one channel specified: " + scopeValue)
            }
          }
        } catch {
          case te: TokenException => throw new MessageException("Invalid Scope: " + scopeValue)
          case cce: ClassCastException => throw new MessageException("Invalid token message for scope validation: " + wholeMessage)
        }
      })
    }
  }

  val ISSUED_TO_CLIENT = new TokenField { def name = "issued_to_client"

    override def required = false

    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      validatePrivileged(this, fieldValue:Option[String], wholeMessage: Message[_])
    }
  }

  val CLIENT_SOURCE_URL = new TokenField { def name = "client_source_url"

    override def required = false

    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateUrl(fieldValue)
      validatePrivileged(this, fieldValue:Option[String], wholeMessage: Message[_])
    }
  }

  val BACKING_GRANTS = new TokenField { def name = "backing_grants"

    override def required = false

    override def validate(fieldValue:Option[String], wholeMessage: Message[_]) {
      validatePrivileged(this, fieldValue:Option[String], wholeMessage: Message[_])
    }
  }

  private def validatePrivileged(field: TokenField, fieldValue: Option[String], wholeMessage: Message[_]) {
    try {
      if (wholeMessage.asInstanceOf[Token].grantType.isPrivileged)
        field.validateRequired(fieldValue)
    } catch {
      case cce: ClassCastException => throw new MessageException("Field " + field.name + " cannot be blank")
    }
  }

}
